text/0000-is-sorted.md

```diff
@@ -0,0 +1,272 @@
+- Feature Name: is_sorted
+- Start Date: 2018-02-24
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add the methods `is_sorted`, `is_sorted_by` and `is_sorted_by_key` to `[T]` and
+`Iterator`.
+
+# Motivation
+[motivation]: #motivation
+
+In quite a few situations, one needs to check whether a sequence of elements
+is sorted. The most important use cases are probably **unit tests** and
+**pre-/post-condition checks**.
+
+The lack of an `is_sorted()` function in Rust's standard library has led to
+[countless programmers implementing their own](https://github.com/search?l=Rust&q=%22fn+is_sorted%22&type=Code&utf8=%E2%9C%93).
+While it is possible to write a one-liner using iterators (e.g.
+`(0..arr.len() - 1).all(|i| arr[i] < arr[i + 1])`), it is still unnecessary
+overhead while writing *and* reading the code.
+
+In [the corresponding issue on the main repository](https://github.com/rust-lang/rust/issues/44370)
+(from which I will reference a few comments) everyone seems to agree on the
+basic premise: we want such a function.
+
+Having `is_sorted()` and friends in the standard library would:
+- prevent people from spending time on writing their own,
+- improve readbility of the code by clearly showing the author's intent,
+- and encourage to write more unit tests and/or pre-/post-condition checks.
+
+Another proof of this functions' usefulness is the inclusion in the
+standard library of many other languages:
+C++'s [`std::is_sorted`](http://en.cppreference.com/w/cpp/algorithm/is_sorted),
+Go's [`sort.IsSorted`](https://golang.org/pkg/sort/#IsSorted),
+D's [`std.algorithm.sorting.is_sorted`](https://dlang.org/library/std/algorithm/sorting/is_sorted.html)
+and others. (Curiously, many (mostly) more high-level programming language –
+like Ruby, Javascript, Java, Haskell and Python – seem to lack such a function.)
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Possible documentation of the three new methods of `Iterator`:
+
+> ```rust
+> fn is_sorted(self) -> bool
+> where
+>     Self::Item: Ord,
```

Comment:
Why `Ord` instead of `PartialOrd`?  Testing if `a <= b <= c <= ... <= z` is a perfectly well-defined operation on a partially-ordered set, and to give floating point numbers the shaft yet again even for something as innocent as this just seems spiteful.

Reply:
Good point, sorry for not properly thinking about that beforehand. So the comparator for the `_by` versions may also return an `Option<Ordering>`, right? And how do we want to handle a `None` then? I might be too sleepy think... but it's not immediately obvious whether `[1.0, NaN, 2.0]` should be considered ordered or not, right?

Reply:
Hm.  This is a predicament.

If this existed in a vacuum, then I would say the comparator should produce `bool`, because an `<=` operation really is most natural definition for a partial ordering. `Option<Ordering>` seems to me to be a sort of an allowance to make deriving Ord from PartialOrd possible.

But obviously, this does not exist in a vacuum, and so a bool comparitor could create unnecessary confusion for users.  So yes, I suppose we should try `Option<Ordering>`.

---

Re: `[1.0, NaN, 2.0]`, I am about to post a much larger comment on this, but in summary, I am of the stance that there is only **one sensible implementation** of `is_sorted` on a partially ordered set, and this implementation returns `false` for a sequence like `[1.0, NaN, 2.0]`.